### DISCUSSION TITLE
(Why) Does or doesn't it make sense to use a state machine instead of a switch/case to manage UI state?

### DISCUSSION THREAD

**Discussion by karfau** [statelyai/xstate#2091] [Q&A] [answered]
Imagine I have to model a sidebar, where there is a flat list of possible states and clicking on any of the items in the list will switch to that state and show the appropriate view in the main area.

I'm wondering: I imagine it would be a very simple switch case statements with a simple list of the possible states (as strings) and maybe a related label. So implementing it as a state machine seems to be a very verbose and inefficient way of doing it especially when doing it in typescript where I need to also define the related Schema and Event types, etc..

Any thoughts?
Any benefits that just don't occur to me right now, but would still make it worth wile?

**Comment by ezhikov** [comment]
I think this article can answer your question: [You don't need a library for state machines](https://dev.to/davidkpiano/you-don-t-need-a-library-for-state-machines-k7h)

In short, you will make state machine anyway, but you not necessarily need XState (or other library) for it. Often small homemade abstraction is more than enough. Still, read the article, it's awesome.

**Reply by karfau** [reply]
Thx for the link, I read through it and liked it, but it didn't fully answer the question for me.
But it after reading it I decided to ["take the conversation to the next level" and add some code to the mix](https://github.com/davidkpiano/xstate/discussions/2091#discussioncomment-611686)

**Comment by karfau** [comment, marked as answer]
I wanted to see what I could come up with to achieve what I have in mind, without using xstate, and make both approaches comparable:
https://codesandbox.io/s/magical-rgb-htz9f?file=/src/index.tsx
(Spoiler: In the end I didn't even use any `switch` but of course I could have.)

**Read thisthread for the actual answer to my question.**

**Reply by davidkpiano** [reply]
This example is too simple to make a reasonable comparison. You're trying to compare XState to what you would normally do to ensure that only one value is allowed at a time. If you're just doing that, I'd recommend the vanilla approach.

**Reply by karfau** [reply]
> This example is too simple to make a reasonable comparison.

I agree, and I'm not comparing this simple scenario to xstate as a whole.

> You're trying to compare XState to what you would normally do to ensure that only one value is allowed at a time.

Yes, I did see a 50+ line implementation (including types) using xstate for doing something very similar to what I did in the example and I was asking myself: "Do we really need to use xstate for this?" which is why I decided to ask this question here.

>  If you're just doing that, I'd recommend the vanilla approach.

Thank you for this statement :)

**Reply by davidkpiano** [reply]
> Yes, I did see a 50+ line implementation (including types) using xstate for doing something very similar to what I did in the example and I was asking myself: "Do we really need to use xstate for this?" which is why I decided to ask this question here.

Looking at it at face-value, it can seem like XState is overkill. But also consider that developers use XState for simple application logic because they _anticipate_ that the logic will get more complex, as features are added etc. So XState can be a good preemptive measure for sound app logic.

Also, have you checked out `@xstate/fsm`? It's a tiny implementation that's perfect for simpler state machines.

**Reply by karfau** [reply]
In that app xstate is already a dependency so the more lightweight version is not needed.

And of course the anticipation of more complex behavior can be a reason to already use a state machine.

But in the concrete example I provided, the difference is so minimal, that it would be very easy to replace `useState` with `useMachine` once it is needed, that I would still argue to stick to `useState` until one knows for sure that it is not enough.

**Reply by davidkpiano** [reply]
Like I mentioned, the example you gave is too trivial; of course `useState` is going to seem like the more straightforward solution.

But even here, there is a potential issue: what if you need to do the steps in order? That is, `one` -> `two` -> `three`? You can handle it with defensive logic:

```js
// clicking button three
if (state === 'two') {
  setState('three')
}
```

But now your logic lives in event handlers instead of a centralized place. And it becomes more difficult to understand the flow of app logic - how will a new developer know how the app is supposed to work? They can decipher all the if-statements (if they can find them) and piece everything together, but that's error-prone and inefficient. And these problems multiply as the app gets bigger.

I encourage you to try a non-trivial example without state machines to understand the value. You can find some at https://xstate-catalogue.com/

**Reply by karfau** [reply]
I fully agree, that if there would be any even forseable case in which at least one state would not be allowed I would switch to a state machine right away.
But there are cases as simple as the one I described where a state machine is not needed.

**Reply by davidkpiano** [reply]
> But there are cases as simple as the one I described where a state machine is not needed.

Agree; if you _know_ your logic is going to be simple, you don't need to use a state machine.

If you _know_ your logic is going to be complex, you should use a state machine.

If you _don't know_ if it's going to be simple or complex, IMO play it safe: use a state machine at the first hint that it might become overly complex, even if it's still simple-ish.

**Comment by davidkpiano** [comment]
Have you read this article? https://dev.to/davidkpiano/no-disabling-a-button-is-not-app-logic-598i

**Reply by karfau** [reply]
It is a very interesting read.
From the title, I first thought it would be an argument for not using a state machine to control the disabled status of a button, but it is actually an argument for not misusing the DOM to control app/UI state and to encapsulate "trivial looking" requirements (that increase during the article) into a state machine.

My code snippet also uses disabling s button to indicate "the current state", but I believe I did it in a way that is in alignment with this article.
